### PR TITLE
[ROCm] Adding rocmprim as a third_party library, minor update to eigen patch

### PR DESCRIPTION
**Pull request by deven-amd**
This PR does two things

1. Add rocmprim as a third_party library, add it to the bazel target(s) which will eventually need it (in PRs to TF operators to be filed soon)

2. Make a minor update to the eigen patch.
The updates are to fix a couple of ROCm/HIP specific errors in the Eigen code. We have already begun the process of upstreaming these changes to Eigen. So this patch update is expected to be temporary in nature.

This PR is pre-requisite to enabling ROCm support for some of TF Operators. Please exepdite it as much as possible.

Thanks

deven

-------------------------------------

@tatianashp , @whchung : just FYI

### CODE REVIEW

**Comment by deven-amd** [comment]
rebased to remove merge conflicts

File: tensorflow/workspace.bzl
Comment (whchung) [inline tensorflow/workspace.bzl]: the package of `rocPRIM` in use LGTM

**Review by whchung — DISMISSED** [review]
LGTM. The PR introduces `rocPRIM` dependency, as well as HIP-specific patch to `Eigen`, which matches what's being used in AMD downstream fork.

**Comment by gunan** [comment]
Adding @rmlarsen for the changes in eigen patch.
Please do not merge without his approval.

**Comment by deven-amd** [comment]
* rebased to remove merge conflict
* also added a commit to fix a compile error that got recently introduced in the `--config=rocm` build

----

@rmlarsen 

For now, I am leaving in the commit for the eigen patch. We can drop that commit from this PR, once the corresponding Eigen PR has been merged + integrated into TF. 

There are a few other PRs that are built on top of this PR, so dropping the eigen commit from this PR (before the TF integration of the Eigen PR) would break the compile for this + all those PRs.

**Comment by deven-amd** [comment]
@rmlarsen 

Thank you for getting the Eigen PR merged. I appreciate your help in expediting that process.

Please update this PR once the TF eigen pointer is updated as well. At that point I will rebase this PR and drop the commit for the eigen patch file update.

Thanks again

deven

**Comment by deven-amd** [comment]
@rmlarsen 

gentle ping.  would you happen to have any insight on when the TF eigen pointer will move foward to include the eigen PR that was merged?

thanks

**Comment by deven-amd** [comment]
@rmlarsen , thank you for updating the TF eigen pointer.

I have re-based this PR to drop the appends to the eigen patch file. 
The only change that remains is the minor change 
from :  `#if defined(EIGEN_CUDA_ARCH)`
to : `#if defined(EIGEN_CUDA_ARCH) || defined(EIGEN_HIP_DEVICE_COMPILE)`
My understanding is that you will upstream this change to Eigen, (if needed). Let me know if that is incorrect.

Please review and approve this PR. It is a pre-req for quite a few other PRs :)

thanks

**Comment by deven-amd** [comment]
@gunan @rmlarsen 

gentle ping

please review and merge the PR. This PR is a gating item for quite a few other ROCm PRs.

thanks

**Comment by deven-amd** [comment]
@gunan @rmlarsen

gentle ping

**Review by angerson — COMMENTED** [review]
The build stuff looks fine to me, but still depends on @rmlarsen's confirmation of the eigen patch.

**Review by rmlarsen — DISMISSED** [review]

**Comment by angersson** [comment]
This change was approved, but is failing internally because third_party dependency changes need special handling for conversion. I'm working on it.

**Comment by angersson** [comment]
@deven-amd I'm running into trouble merging this due to differences in internal behavior for `is_rocm/cuda` and `is_rocm/cuda_configured`. Can you explain why you needed to make changes with the `cuda` calls? Favoring `is_rocm` where `is_cuda` is used (and similar for where `is_cuda_configured`) would probably resolve these issues.

**Comment by deven-amd** [comment]
@angersson 

@chsigg had a similar query in another PR...please see the discussion here (https://github.com/tensorflow/tensorflow/pull/26753#discussion_r267022934) for details.

The summary of it is that, the change from `if_cuda / if_rocm` to `if_cuda_is_configured / if_rocm_is_configured` is needed in scenarios where there is a common target being added by by both (`rocm` and `cuda`) to the dependency list. 

Using `if_cuda / if_rocm` (which returns a select statement) in such scenarios results in a bazel error (duplicate depedency). 

Switching to `if_cuda_is_configured / if_rocm_is_configured` (which returns the `true` arg or empty list) resolves the bazel error.

Hope this clears things up.

**Comment by angersson** [comment]
@chsigg Can you look at importing this, or find someone who can, please? Thanks!

**Review by chsigg — DISMISSED** [review]
The PR looks ok, but I would like to eventually come up with a better solution to our if_*() vs if_*_is_configured(). Maybe adding an `if_gpu(if_cuda_or_rocm, otherwise)` is the right way to go.

**Comment by deven-amd** [comment]
@chsigg   gentle ping

how close are we to getting this PR merged?

thanks

**Review by ezhulenev — DISMISSED** [review]

**Comment by deven-amd** [comment]
the CI failures seem unrelated to the changes in this PR....not sure what is going on

**Comment by ezhulenev** [comment]
This is a problem with a internal/external build setup mismatch. I'll try to merge it tomorrow, but probably I'll have to find someone who understands our build tooling better than me... might take few days.

**Comment by ezhulenev** [comment]
I've merged non-build related changes, could you please sync it with master, I'll try to find someone who can help with build-related changes.

**Comment by deven-amd** [comment]
@ezhulenev , thanks for merging the non-build commits.

I will rebase this branch onto master, and push it out later today. After the update, there will be only one commit left in this PR (build change for rocprim)

**Comment by deven-amd** [comment]
@ezhulenev rebase done

**Comment by gunan** [comment]
I am trying to merge this, but this is proving to be very very difficult.
I had to revert all the changes from "if_cuda" to "if_cuda_is_configured"
Next time, let's split such changes into two, where one only adds the new external dependency, and another where we start using it.

**Comment by deven-amd** [comment]
@gunan thank you for getting this merge done

will keep your suggestions in mind, the next time we need to add a dependency.

**Comment by deven-amd** [comment]
@gunan 

> I had to revert all the changes from "if_cuda" to "if_cuda_is_configured"

The change from `if_cuda` to `if_cuda_is_configured` was intentional and needed for the `--config=rocm` build to work. With that change reverted, the `--config=rocm` build is broken :(

How can I help you resolve this such that `--config=rocm` build is working again?

@chsigg, just FYI to keep you in the loop

thanks

deven

----------------

build errors I get when trying to do a `--config=rocm` build
```
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4029:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'bias_op_gpu'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4029:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'bias_op'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4029:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'libtfkernel_bias_op.so'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4069:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'softmax_op_gpu'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4069:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'softmax_op'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4069:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'libtfkernel_softmax_op.so'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4735:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'sparse_xent_op_gpu'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4735:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'sparse_xent_op'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:4735:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'libtfkernel_sparse_xent_op.so'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:5227:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'multinomial_op_gpu'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:5227:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'multinomial_op'
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:5227:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'libtfkernel_multinomial_op.so'
ERROR: /root/tensorflow/tensorflow/core/BUILD:1558:1: Target '//tensorflow/core/kernels:dataset_ops' contains an error and its package is in error and referenced by '//tensorflow/core:all_kernels'
ERROR: /root/tensorflow/tensorflow/core/BUILD:1558:1: Target '//tensorflow/core/kernels:list_kernels' contains an error and its package is in error and referenced by '//tensorflow/core:all_kernels'
ERROR: Analysis of target '//tensorflow/tools/pip_package:build_pip_package' failed; build aborted: Cannot compute config conditions
INFO: Elapsed time: 13.208s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (366 packages loaded, 10066 targets configured)
```

**Pull request merged**
